proc/proc.go

```diff
@@ -146,69 +144,45 @@ func (dbp *Process) LoadInformation(path string) error {
 	return nil
 }
 
-// Find a location by string (file+line, function, breakpoint id, addr)
-func (dbp *Process) FindLocation(str string) (uint64, error) {
-	// File + Line
-	if strings.ContainsRune(str, ':') {
-		fl := strings.Split(str, ":")
-
-		fileName, err := filepath.Abs(fl[0])
-		if err != nil {
-			return 0, err
-		}
-
-		line, err := strconv.Atoi(fl[1])
-		if err != nil {
-			return 0, err
-		}
+func (dbp *Process) FindFileLocation(fileName string, lineno int) (uint64, error) {
+	pc, _, err := dbp.goSymTable.LineToPC(fileName, lineno)
+	if err != nil {
+		return 0, err
+	}
+	return pc, nil
+}
 
-		pc, _, err := dbp.goSymTable.LineToPC(fileName, line)
-		if err != nil {
-			return 0, err
-		}
-		return pc, nil
+// Finds address of a function's line
+// If firstLine == true is passed FindFunctionLocation will attempt to find the first line of the function
+// If lineOffset is passed FindFunctionLocation will return the address of that line
+// Pass lineOffset == 0 and firstLine == false if you want the address for the function's entry point
+// note that setting breakpoints at that address will cause surpirsing behavior:
+// https://github.com/derekparker/delve/issues/170
+func (dbp *Process) FindFunctionLocation(funcName string, firstLine bool, lineOffset int) (uint64, error) {
+	fn := dbp.goSymTable.LookupFunc(funcName)
+	if fn == nil {
+		return 0, fmt.Errorf("Could not find function %s\n", funcName)
 	}
 
-	// Try to lookup by function name
-	fn := dbp.goSymTable.LookupFunc(str)
-	if fn != nil {
-		// fn.Entry, the entry point of the function, is always a prologue, the prologue may call into the scheduler or grow the stack,
-		// this will result in the breakpoint getting hit multiple times without any apparent program progress inbetween.
-		// In order to avoid this confusing behaviour we try to find the first line of the function and set the breakpoint there.
+	if firstLine {
 		filename, lineno, _ := dbp.goSymTable.PCToLine(fn.Entry)
-		var firstLinePC uint64
-		var err error
-		for {
-			lineno++
-			firstLinePC, _, err = dbp.goSymTable.LineToPC(filename, lineno)
-			if err == nil {
-				break
-			}
-			if _, unk := err.(*gosym.UnknownLineError); !unk {
-				return 0, err
-			}
+		lines, err := dbp.ast.NextLines(filename, lineno)
```

Comment:
This function explicitly only supports Go source files, what happens if the user specifies a cgo func?